### ISSUE TITLE
Dlib Anaconda fps face detection bottleneck

### ISSUE THREAD

**Issue by kdx2** [davisking/dlib#1556]
Hello community,
I am trying to carry out facial detection on 1080p 60fps video files. For this purpose I am using dlib, especially the `get_frontal_face_detector()` and `shape_predictor()` which I use to find the 68 landmarks. 

I am currently doing `detector(frame, 0)`

After some exhaustive measurements I realised that the dlib face detector is consuming more than 90% of the time of the program execution, processing a 2 or 3-second-long videos. The even worse bit is that instead of being close to 60fps I am now at around 1.8 - 2fps in the processing and I need speedup.

I have tried for instance: dlib.DLIB_USE_CUDA and it says `False`. I was wondering that if there is a way to enable that it may give me a significant speed-up in the fps. Do you have any idea how I can do that in Anaconda, as it is the one installing and compiling libraries for me?  

When doing:
`import dlib.cuda as cuda`
`> No module named 'dlib.cuda'; 'dlib' is not a package`

I honestly, do want to stick to Anaconda, instead of making global installations. Any optimization or idea would be highly valuable.

Thank you in advance,

Dlib version: 19.9.0
The copy of dlib is from the anaconda conda-forge menpo repository.
Windows 64-bit.
Anaconda 1.6

**Comment by kdx2** [comment]
An update on where I am currently at:
I managed to turn on the AVX and SSE4 instructions, and again failed at enabling the cuda support. Here comes the log:
There are a couple things I don't understand.
1. Why does it see Cuda v8.0 when I just uninstalled all v8.0 drivers and stuff and made a clean install of v9.0. I have also checked all my environmental variables and made sure that they are v9.0 not 8.
2. I did see this [issue]https://github.com/davisking/dlib/issues/1035(url) which is the only closest to mine but even though I made all described [here]https://github.com/davisking/dlib/issues/1035#issuecomment-435780705(url) seems build tool does not like the VS. From all done by now, the optimization AVX and SSE4 give me is +0.5fps of speed up. I desperately need that GPU and any other suggestions which you may have.
3. Where should I run that given that I am installing with setup.py? cmake .. -G "Visual Studio 14 2015 Win64" -T host=x64 

Here comes the 
`>python setup.py install --yes USE_AVX_INSTRUCTIONS --yes DLIB_USE_CUDA
running install
running bdist_egg
running egg_info
writing dlib.egg-info\PKG-INFO
writing top-level names to dlib.egg-info\top_level.txt
writing dependency_links to dlib.egg-info\dependency_links.txt
package init file 'dlib\__init__.py' not found (or not a regular file)
reading manifest file 'dlib.egg-info\SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'dlib.egg-info\SOURCES.txt'
installing library code to build\bdist.win-amd64\egg
running install_lib
running build_py
running build_ext
Building extension for Python 3.5.6 |Anaconda custom (64-bit)| (default, Aug 26 2018, 16:05:27) [MSC v.1900 64 bit (AMD64)]
Invoking CMake setup: 'cmake C:\Users\konyd\dlib-master\tools\python -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\Users\konyd\dlib-master\build\lib.win-amd64-3.5 -DPYTHON_EXECUTABLE=D:\Software\Anaconda\envs\tensorflow\python.exe -DUSE_AVX_INSTRUCTIONS=yes -DDLIB_USE_CUDA=yes -DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=C:\Users\konyd\dlib-master\build\lib.win-amd64-3.5 -A x64'
-- The C compiler identification is MSVC 19.14.26433.0
-- The CXX compiler identification is MSVC 19.14.26433.0
-- Check for working C compiler: D:/Software/Visual Studio Communty/VC/Tools/MSVC/14.14.26428/bin/Hostx86/x64/cl.exe
-- Check for working C compiler: D:/Software/Visual Studio Communty/VC/Tools/MSVC/14.14.26428/bin/Hostx86/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: D:/Software/Visual Studio Communty/VC/Tools/MSVC/14.14.26428/bin/Hostx86/x64/cl.exe
-- Check for working CXX compiler: D:/Software/Visual Studio Communty/VC/Tools/MSVC/14.14.26428/bin/Hostx86/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: D:/Software/Anaconda/envs/tensorflow/python.exe (found version "3.5.6")
-- pybind11 v2.2.2
-- Using CMake version: 3.12.4
-- Compiling dlib version: 19.16.99
-- SSE4 instructions can be executed by the host processor.
-- AVX instructions can be executed by the host processor.
-- Enabling AVX instructions
-- Found CUDA: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v8.0 (found suitable version "8.0", minimum required is "7.5")
-- Looking for cuDNN install...
-- Found cuDNN: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v8.0/lib/x64/cudnn.lib
-- Disabling CUDA support for dlib.  DLIB WILL NOT USE CUDA
-- C++11 activated.
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/konyd/dlib-master/build/temp.win-amd64-3.5/Release
Invoking CMake build: 'cmake --build . --config Release -- /m'
Microsoft (R) Build Engine version 15.7.180.61344 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 18/11/2018 23:06:39.
     1>Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\ALL_BUILD.vcxproj" on node 1 (default targets).
     1>Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\ALL_BUILD.vcxproj" (1) is building "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
     2>InitializeBuildStatus:
         Creating "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
       CustomBuild:
         All outputs are up-to-date.
       FinalizeBuildStatus:
         Deleting file "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
         Touching "x64\Release\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
     2>Done Building Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\ZERO_CHECK.vcxproj" (default targets).
     1>Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\ALL_BUILD.vcxproj" (1) is building "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\dlib_build\dlib.vcxproj" (3) on node 2 (default targets).
     3>InitializeBuildStatus:
         Creating "dlib.dir\Release\dlib.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
       CustomBuild:
         All outputs are up-to-date.
       ClCompile:
         All outputs are up-to-date.
         All outputs are up-to-date.
       Lib:
         All outputs are up-to-date.
         dlib.vcxproj -> C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\dlib_build\Release\dlib19.16.99_release_64bit_msvc1914.lib
       FinalizeBuildStatus:
         Deleting file "dlib.dir\Release\dlib.tlog\unsuccessfulbuild".
         Touching "dlib.dir\Release\dlib.tlog\dlib.lastbuildstate".
     3>Done Building Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\dlib_build\dlib.vcxproj" (default targets).
     1>Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\ALL_BUILD.vcxproj" (1) is building "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\dlib_python.vcxproj" (4) on node 3 (default targets).
     4>InitializeBuildStatus:
         Creating "dlib_python.dir\Release\dlib_python.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
       CustomBuild:
         All outputs are up-to-date.
       ClCompile:
         All outputs are up-to-date.
       Link:
         All outputs are up-to-date.
         dlib_python.vcxproj -> C:\Users\konyd\dlib-master\build\lib.win-amd64-3.5\dlib.cp35-win_amd64.pyd
       FinalizeBuildStatus:
         Deleting file "dlib_python.dir\Release\dlib_python.tlog\unsuccessfulbuild".
         Touching "dlib_python.dir\Release\dlib_python.tlog\dlib_python.lastbuildstate".
     4>Done Building Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\dlib_python.vcxproj" (default targets).
     1>InitializeBuildStatus:
         Creating "x64\Release\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
       CustomBuild:
         All outputs are up-to-date.
       FinalizeBuildStatus:
         Deleting file "x64\Release\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild".
         Touching "x64\Release\ALL_BUILD\ALL_BUILD.tlog\ALL_BUILD.lastbuildstate".
     1>Done Building Project "C:\Users\konyd\dlib-master\build\temp.win-amd64-3.5\Release\ALL_BUILD.vcxproj" (default targets).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.78
creating build\bdist.win-amd64\egg
copying build\lib.win-amd64-3.5\dlib.cp35-win_amd64.pyd -> build\bdist.win-amd64\egg
creating stub loader for dlib.cp35-win_amd64.pyd
byte-compiling build\bdist.win-amd64\egg\dlib.py to dlib.cpython-35.pyc
creating build\bdist.win-amd64\egg\EGG-INFO
copying dlib.egg-info\PKG-INFO -> build\bdist.win-amd64\egg\EGG-INFO
copying dlib.egg-info\SOURCES.txt -> build\bdist.win-amd64\egg\EGG-INFO
copying dlib.egg-info\dependency_links.txt -> build\bdist.win-amd64\egg\EGG-INFO
copying dlib.egg-info\not-zip-safe -> build\bdist.win-amd64\egg\EGG-INFO
copying dlib.egg-info\top_level.txt -> build\bdist.win-amd64\egg\EGG-INFO
writing build\bdist.win-amd64\egg\EGG-INFO\native_libs.txt
creating 'dist\dlib-19.16.99-py3.5-win-amd64.egg' and adding 'build\bdist.win-amd64\egg' to it
removing 'build\bdist.win-amd64\egg' (and everything under it)
Processing dlib-19.16.99-py3.5-win-amd64.egg
creating d:\software\anaconda\envs\tensorflow\lib\site-packages\dlib-19.16.99-py3.5-win-amd64.egg
Extracting dlib-19.16.99-py3.5-win-amd64.egg to d:\software\anaconda\envs\tensorflow\lib\site-packages
Adding dlib 19.16.99 to easy-install.pth file

Installed d:\software\anaconda\envs\tensorflow\lib\site-packages\dlib-19.16.99-py3.5-win-amd64.egg
Processing dependencies for dlib==19.16.99
Finished processing dependencies for dlib==19.16.99`

**Comment by kdx2** [comment]
Hello Mr. King,
I did not want to publish code because it is part of my university project
and if I publish that before the marking then I'll have to reference my own
code (+ on all other websites if any other issues arise in the future). So,
I'm trying to be a bit discrete. Thank you a lot for the feedback,
especially about the bit saying that cuda won't do anything, this will save
me time.
I am doing facial microexpressions recognition, which last *most up to*
*0.5s* in duration, so I do need to process every frame without skipping, I
believe. This is how the code looks like:

detectFaceFrame = dl.get_frontal_face_detector() # somewhere up the code

while(camera.isOpened()):

    isCameraWorking, currentFrame = camera.read()
    if isCameraWorking is True:
        # Color frame to Grayscale
        currentFrameGray = cv.cvtColor(currentFrame, cv.COLOR_BGR2GRAY)
        # 0 means, do not do any upsampling of the image.
        *allFramedFaces = detectFaceFrame(currentFrame, 0)*

        for (i, singleFramedFace) in enumerate(allFramedFaces): # usually
there will be just one person in facing the camera so this loop in general
will run only once
*            landmarksDetected = predictLandmarkPoints(currentFrame,
singleFramedFace)*
            listXYPairs = shape_to_np(landmarksDetected)
            xUp, yUp, xBottom, yBottom, centroid =
findFaceFrame(currentFrameGray,

listXYPairs)
... Then other things follow but really what throttles the whole system
like crazy is the detectFaceFrame(currentFrame, 0) that I call, it takes
91% of the execution time alone. This is ran on a 3s video clip. The
processing takes 60s => around 54s are spent into facedetection. The idea
behind face detection and landmarking is: I find where the face is, so I
can find where the middle of the nose is. When I find the middle of the
nose, I can build the rectangle of a fixed size (very important) with the
face perfectly centered. Then that face is cropped and its image - sent to
an LRCN for recognition.

If you have any idea of how I can boost the face detection stage, I will be
super happy. Haar from OpenCV didn't give me much better performance,
either. Not to mention, it was with more false positives. Maybe I am doing
something wrong. Maybe the frontal face detector is not the way to go. If
you have any suggestions, please, share them with me. Tomorrow I will
attempt to apply the downsampling.
